core/validator_manager.go

```diff
@@ -0,0 +1,151 @@
+
+	// quorumSize represents quorum for the height specified in the current View
+	quorumSize *big.Int
+
+	// validatorsVotingPower is a map of the validator addresses on their voting power for
+	// the height specified in the current View
+	validatorsVotingPower map[string]*big.Int
+
+	backend ValidatorBackend
+
+	log Logger
+}
+
+// NewValidatorManager creates new ValidatorManager
+func NewValidatorManager(backend ValidatorBackend, log Logger) *ValidatorManager {
+	return &ValidatorManager{
+		quorumSize:            big.NewInt(0),
+		backend:               backend,
+		validatorsVotingPower: nil,
+		log:                   log,
+		vpLock:                &sync.RWMutex{},
+	}
+}
+
+// Init sets voting power and quorum size
+func (vm *ValidatorManager) Init(height uint64) error {
+	vm.vpLock.Lock()
+	defer vm.vpLock.Unlock()
+
+	validatorsVotingPower, err := vm.backend.GetVotingPowers(height)
+	if err != nil {
+		return err
+	}
+
+	totalVotingPower := calculateTotalVotingPower(validatorsVotingPower)
+	if totalVotingPower.Cmp(big.NewInt(0)) <= 0 {
+		return errVotingPowerNotCorrect
+	}
+
+	vm.validatorsVotingPower = validatorsVotingPower
+	vm.quorumSize = calculateQuorum(totalVotingPower)
+
+	return nil
+}
+
+// HasQuorum provides information on whether messages have reached the quorum
+func (vm *ValidatorManager) HasQuorum(addressMap map[string]struct{}) bool {
+	vm.vpLock.RLock()
+	defer vm.vpLock.RUnlock()
+
+	// if not initialized correctly return false
+	if vm.validatorsVotingPower == nil {
+		return false
+	}
+
+	messageVotePower := big.NewInt(0)
+
+	for from := range addressMap {
+		if vote, ok := vm.validatorsVotingPower[from]; ok {
+			messageVotePower.Add(messageVotePower, vote)
+		}
+	}
+
+	return messageVotePower.Cmp(vm.quorumSize) >= 0
+}
+
+// HasPrepareQuorum provides information on whether prepared messages have reached the quorum
+func (vm *ValidatorManager) HasPrepareQuorum(proposalMessage *proto.Message, msgs []*proto.Message) bool {
+	if proposalMessage == nil {
+		vm.log.Info("HasPrepareQuorum - proposalMessage is not set")
+
+		return false
+	}
+
+	proposerAddress := proposalMessage.From
+	sendersAddressesMap := map[string]struct{}{
+		string(proposerAddress): {},
+	}
+
+	for _, message := range msgs {
+		if bytes.Equal(message.From, proposerAddress) {
+			vm.log.Info("HasPrepareQuorum - proposer is among signers but it is not expected to be")
+
+			return false
+		}
+
+		sendersAddressesMap[string(message.From)] = struct{}{}
+	}
+
+	return vm.HasQuorum(sendersAddressesMap)
+}
+
+func calculateQuorum(totalVotingPower *big.Int) *big.Int {
+	quorum := new(big.Int).Mul(totalVotingPower, big.NewInt(2))
+
+	return bigIntDivCeil(quorum, big.NewInt(3))
+}
+
+func calculateTotalVotingPower(validatorsVotingPower map[string]*big.Int) *big.Int {
+	totalVotingPower := big.NewInt(0)
+	for _, validatorVotingPower := range validatorsVotingPower {
+		totalVotingPower = totalVotingPower.Add(totalVotingPower, validatorVotingPower)
```

Comment:
Optional nitpick:
`func (*big.Int) Add` function is a strange one, as it both mutates instances being called against and returns a result.

So it would be ok invoking this way as well:
```suggestion
		totalVotingPower.Add(totalVotingPower, validatorVotingPower)
```

Reply:
fixed https://github.com/0xPolygon/go-ibft/pull/72/commits/cb9e07e760de1d26add8c94d1ba82a6490e73f4c